src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -285,7 +285,7 @@ void primitive_inst::update_shape() {
         has_runtime_deps = true;
     }
 
-    if (has_runtime_deps) {
+    if (has_runtime_deps && !get_node().is_in_shape_of_subgraph()) {
```

Comment:
Not sure that we can really have such condition here. What if we have GPU impl in shape_of subgraph?

Reply:
@vladimir-paramuzov currently we use only CPU implementations in subgraphs, so this optimization looks acceptable to me

Reply:
@sshlyapn That doesn't make me more confident in this change. What if a new model that we haven't checked yet has new op on the shape-of subgraph w/o cpu impl? I expect that in such case we'll face sporadic accuracy issue which will be hard to troubleshoot.

Reply:
@vladimir-paramuzov  Hmm, yeah, you are right, it makes sense in case of complex patterns when single shape-of subgraph is splitted by some new operation like this:
![image](https://github.com/openvinotoolkit/openvino/assets/3998162/0e300b03-feec-45a6-b687-12e0ed2844c4)

I think we may additionally check if runtime dependency is in shape-of subgraph and then do not set `has_runtime_deps`. Seems it will be enough to keep improvements from this patch and avoid such problematic cases and , because as of now only nodes with cpu impls may be included to shape-of subgraph.

https://github.com/openvinotoolkit/openvino/blob/52b9df4a6d7cf80833c0e441c11fb80f12d6bfce/src/plugins/intel_gpu/src/graph/graph_optimizer/mark_shape_of_subgraphs.cpp#L59-L65

Reply:
Thanks for the discussion! Or, how about not to insert the clFinish if the predecessor's impl is cpu ?

Reply:
Of course the above suggestion can be applied at the same time too.

Reply:
> Or, how about not to insert the clFinish just if the runtime dep predecessor's impl is cpu ?

It could be possible, but please note that such solution may be impacted by some changes in the plugin, e.g. if we implement async execution of cpu impls or use different impl types for static and dynamic kernels

Reply:
Updated the PR.